src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/EnumerableExtensions.cs

```diff
@@ -38,6 +49,49 @@ static ImmutableArray<TResult> BuildResult(IEnumerable<T> items, Func<T, TResult
         }
     }
 
+    /// <summary>
+    ///  Projects each element of an <see cref="IEnumerable{T}"/> into a new form by incorporating the element's index.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements in <paramref name="source"/>.</typeparam>
+    /// <typeparam name="TResult">The type of the value returned by <paramref name="selector"/>.</typeparam>
+    /// <param name="source">A sequence of values to invoke a transform function on.</param>
+    /// <param name="selector">
+    ///  A transform function to apply to each source element; the second parameter of
+    ///  the function represents the index of the source element.
+    /// </param>
+    /// <returns>
+    ///  Returns a new <see cref="ImmutableArray{T}"/> whose elements are the result of invoking the transform function
+    ///  on each element of <paramref name="source"/>.
+    /// </returns>
+    public static ImmutableArray<TResult> SelectAsArray<T, TResult>(this IEnumerable<T> source, Func<T, int, TResult> selector)
+    {
+        if (source is ImmutableArray<T> array)
+        {
+            return ImmutableArrayExtensions.SelectAsArray(array, selector);
+        }
+
+        if (source is IReadOnlyList<T> list)
+        {
+            return ReadOnlyListExtensions.SelectAsArray(list, selector);
+        }
+
+        return BuildResult(source, selector);
+
+        static ImmutableArray<TResult> BuildResult(IEnumerable<T> items, Func<T, int, TResult> selector)
+        {
+            using var results = new PooledArrayBuilder<TResult>();
```

Comment:
Is it worth using `TryGetCount()` to set a capacity? Or would any case that helps with be covered by the two ifs above?

Reply:
Great suggestion! `TryGetCount()` is definitely useful here. It tests for `IReadOnlyCollection<T>`, `ICollection<T>` and `ICollection` -- plus iterators with a known size on .NET Core. None of those are covered by the cases above.